### PR TITLE
Parser Optimization, Performance Enhancements & Code Cleanup

### PR DESCRIPTION
**Pull request by vagetman**
# 🛠️ Parser and Evaluation Optimization

## 🚀 Overview
This PR refactors the expression evaluation and parsing logic to improve **performance, maintainability, and correctness**. The main changes include **removing unnecessary allocations**,  and **optimizing string handling**.

---

### Optimized String Handling
- Replaced `String` in `Value::Text` with `Cow<'static, str>` to **reduce unnecessary allocations** and improve efficiency by avoiding cloning where possible.
- Updated `EvalContext` to use `Cow` for **efficient storage and retrieval of variables**, ensuring that string data is managed more effectively.

### Improved Expression Evaluation (`eval_expr`)
- Eliminated redundant `.to_string()` conversions, which were causing unnecessary string allocations.
- Ensured efficient variable resolution by using **borrowed references instead of cloned strings**, reducing memory usage.

### Enhanced Error Messages & Documentation
- Improved error messages for better **debugging and user feedback**. 
- Added comments to clarify logic in key parsing functions, enhancing **code readability and maintainability**.

### Testing and Validation
- **Expanded Test Coverage**: Introduced new test cases to cover previously untested scenarios, ensuring robustness and catching edge cases.
- **Bug Fixes**: Addressed known issues and edge cases that were causing unexpected behavior in certain conditions, improving reliability.

### Code Refactoring & Cleanup
- **Simplified Logic**: Refactored complex conditional statements and loops for better readability, efficiency, and to reduce potential for errors.
- **Consistent Formatting**: Applied uniform code formatting and styling to **enhance code consistency and maintainability**.

### Performance Enhancements
- **Optimized Parsing Logic**: Improved the ESI parsing mechanism to handle edge cases more efficiently, reducing processing time.
- **Reduced Memory Footprint**: Implemented more efficient data structures, leading to lower memory consumption and better performance in memory-constrained environments

## 📈 Impact
✅ **Better Performance** – Fewer memory allocations, faster evaluation, and optimized parsing logic.  
✅ **Lower Memory Usage** – Reduced heap allocations using `Cow<'static, str>` and optimized data structures.  
✅ **Simplified Code** – Refactored complex logic, and improved formatting for better readability.  
✅ **Improved Maintainability** – Cleaner parsing logic, better error messages, and enhanced documentation.

### CODE REVIEW

File: esi/src/expression.rs [hunk 1]
```diff
@@ -128,52 +149,74 @@ impl EvalContext {
...
+/// - `Integer`: A 32-bit signed integer
+/// - `String`: A UTF-8 string
+/// - `Boolean`: A boolean value (true/false)
+/// - `Null`: Represents an absence of value
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub enum Value {
     Integer(i32),
-    String(String),
-    Boolean(BoolValue),
+    Text(Cow<'static, str>),
```
Comment (tyler) [inline esi/src/expression.rs:172]: Why the rename of String to Text?

**Review by tyler — COMMENTED** [review]
Thanks for this :)

File: esi/src/expression.rs (hunk 1 above)
Comment (tyler) [inline esi/src/expression.rs:172]: Oh, also why not convert everything to `&str`? Just trying to avoid all the lifetimes?

**Review by kailan — APPROVED** [review]
This looks great! Thanks for the extra documentation too.

File: esi/src/expression.rs (hunk 1 above)
Comment (vagetman) [reply esi/src/expression.rs:172]: The `String` is  renamed to `Text` because it's `(Cow<'static, str>)` and not `String`... I think it's just an arbitrary name. Naming is hard :)

Comment (vagetman) [reply esi/src/expression.rs:172]: @tyler I'm going to resolve this to get it merged. Please open another PR if you want to improve it.
